sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,477 @@
+additional kernels without involving the host program. In OpenCL 2.0 and beyond,
+kernel invocations can be initiated without host program interaction through
+device side enqueue. C++11 also introduced `std::async()` for task-level
+parallelism. However, neither of these approaches is a good fit for spatial
+compute platforms, which require a clear mapping of parallel constructs to
+spatial instances in the hardware.
+
+Rather than significantly changing the semantics of existing solutions, this
+extension introduces a sub-kernel task-level parallelism interface to SYCL for
+use in device code. It provides the class
+`sycl::ext::intel::experimental::task_sequence`, which is an
+abstraction of an asynchronous sequence of invocations of a callable that can be
+executed asynchronously from the caller, and with which a clear mapping to
+spatial hardware instances can be reasoned about when developing applications.
+
+NOTE: In this document, we use `task_sequence` to indicate the proposed
+`sycl::ext::intel::experimental::task_sequence`
+
+== Definitions
+
+== `task_sequence` class
+
+[source,c++,linenums]
+----
+namespace sycl {
+namespace ext {
+namespace intel {
+namespace experimental {
+
+template<auto& f, class propertiesT =
+   decltype(oneapi::experimental::properties{})>
+class task_sequence{};
+
+template <typename ReturnT, typename... ArgsT, ReturnT (&f)(ArgsT...),
+          class propertiesT>
+class task_sequence<f, propertiesT> {
+
+public:
+  task_sequence(const task_sequence &) = delete;
+  task_sequence &operator=(const task_sequence &) = delete;
+  task_sequence(task_sequence &&) = delete;
+  task_sequence &operator=(task_sequence &&) = delete;
+
+  task_sequence();
+
+  task_sequence(const propertiesT &);
+
+  void async(ArgsT... args);
+
+  ReturnT get();
+  
+  template <typename propertyT>
+  static constexpr bool has_property();
+  
+  template <typename propertyT>
+  static constexpr auto get_property();
+
+  ~task_sequence();
+};
+
+} // namespace experimental
+} // namespace intel
+} // namespace ext
+} // namespace sycl
+
+----
+
+`task_sequence` is a class template, parameterized by an `auto` reference to a
++Callable+ `f` and by a list of properties `propertiesT`.
+The +Callable+ `f` defines the asynchronous task to be
+associated with the `task_sequence`, and requiring an auto reference ensures
+that each `f` be statically resolvable at compile time.  Static resolvability by
+the compiler is desirable when compiling for spatial architectures as it can
+enable the generation of more efficient hardware. Furthermore, the partial
+specialization `typename ReturnT, typename ... ArgsT, ReturnT(&f) (ArgsT...)`
+ensures that the return type `ReturnT` of `f` and the argument types
+`typename ... ArgsT` of `f` are resolvable and fixed for each definition of
+`task_sequence`. 
+
+The `task_sequence` interface consists of two methods, `async` and `get`. The
```

Comment:
Can we make sure the following is clearly stated in the spec?/

By calling async on a task_sequence more than once, the user implies that the invocations of f can be run in parallel.

Reply:
By the way, in C++, the term *member function* is used instead of *method*.

Reply:
Updated both, thanks :)